homeassistant/components/xiaomi_miio/config_flow.py

```diff
@@ -237,7 +237,9 @@ async def async_step_cloud(
                     step_id="cloud", data_schema=DEVICE_CLOUD_CONFIG, errors=errors
                 )
 
-            miio_cloud = MiCloud(cloud_username, cloud_password)
+            miio_cloud = await self.hass.async_add_executor_job(
+                MiCloud, cloud_username, cloud_password
```

Comment:
The call/constructor doesn't do any I/O: https://github.com/Squachen/micloud/blob/master/micloud/micloud.py#L24-L43

IMO there is no need for doing this async. Do I miss something?

Reply:
tzlocal does file operations when trying to determine the hosts timezone

Reply:
There are 3 blocking I/O operations during instantiation.

```shell
2024-09-27 08:23:02.481 WARNING (MainThread) [homeassistant.util.loop] Detected blocking call to open with args ('/etc/timezone',) inside the event loop by integration 'xiaomi_miio' at homeassistant/components/xiaomi_miio/config_flow.py, line 240: miio_cloud = MiCloud(cloud_username, cloud_password) (offender: /home/vscode/.local/ha-venv/lib/python3.12/site-packages/tzlocal/unix.py, line 62: with open(tzpath) as tzfile:), please create a bug report at https://github.com/home-assistant/core/issues?q=is%3Aopen+is%3Aissue+label%3A%22integration%3A+xiaomi_miio%22
For developers, please see https://developers.home-assistant.io/docs/asyncio_blocking_operations/#open
Traceback (most recent call last):
  File "/home/vscode/.local/ha-venv/bin/hass", line 8, in <module>
    sys.exit(main())
  File "/workspaces/ha-core/homeassistant/__main__.py", line 209, in main
    exit_code = runner.run(runtime_conf)
  File "/workspaces/ha-core/homeassistant/runner.py", line 189, in run
    return loop.run_until_complete(setup_and_run_hass(runtime_config))
  File "/usr/local/lib/python3.12/asyncio/base_events.py", line 674, in run_until_complete
    self.run_forever()
  File "/usr/local/lib/python3.12/asyncio/base_events.py", line 641, in run_forever
    self._run_once()
  File "/usr/local/lib/python3.12/asyncio/base_events.py", line 1979, in _run_once
    handle._run()
  File "/usr/local/lib/python3.12/asyncio/events.py", line 88, in _run
    self._context.run(self._callback, *self._args)
  File "/home/vscode/.local/ha-venv/lib/python3.12/site-packages/aiohttp/web_protocol.py", line 477, in _handle_request
    resp = await request_handler(request)
  File "/home/vscode/.local/ha-venv/lib/python3.12/site-packages/aiohttp/web_app.py", line 559, in _handle
    return await handler(request)
  File "/home/vscode/.local/ha-venv/lib/python3.12/site-packages/aiohttp/web_middlewares.py", line 117, in impl
    return await handler(request)
  File "/workspaces/ha-core/homeassistant/components/http/security_filter.py", line 92, in security_filter_middleware
    return await handler(request)
  File "/workspaces/ha-core/homeassistant/components/http/forwarded.py", line 83, in forwarded_middleware
    return await handler(request)
  File "/workspaces/ha-core/homeassistant/components/http/request_context.py", line 26, in request_context_middleware
    return await handler(request)
  File "/workspaces/ha-core/homeassistant/components/http/ban.py", line 85, in ban_middleware
    return await handler(request)
  File "/workspaces/ha-core/homeassistant/components/http/auth.py", line 242, in auth_middleware
    return await handler(request)
  File "/workspaces/ha-core/homeassistant/components/http/headers.py", line 32, in headers_middleware
    response = await handler(request)
  File "/workspaces/ha-core/homeassistant/helpers/http.py", line 73, in handle
    result = await handler(request, **request.match_info)
  File "/workspaces/ha-core/homeassistant/components/http/decorators.py", line 81, in with_admin
    return await func(self, request, *args, **kwargs)
  File "/workspaces/ha-core/homeassistant/components/config/config_entries.py", line 222, in post
    return await super().post(request, flow_id)
  File "/workspaces/ha-core/homeassistant/components/http/data_validator.py", line 74, in wrapper
    return await method(view, request, data, *args, **kwargs)
  File "/workspaces/ha-core/homeassistant/helpers/data_entry_flow.py", line 122, in post
    result = await self._flow_mgr.async_configure(flow_id, data)
  File "/workspaces/ha-core/homeassistant/data_entry_flow.py", line 370, in async_configure
    result = await self._async_configure(flow_id, user_input)
  File "/workspaces/ha-core/homeassistant/data_entry_flow.py", line 417, in _async_configure
    result = await self._async_handle_step(
  File "/workspaces/ha-core/homeassistant/data_entry_flow.py", line 520, in _async_handle_step
    result: _FlowResultT = await getattr(flow, method)(user_input)
  File "/workspaces/ha-core/homeassistant/components/xiaomi_miio/config_flow.py", line 240, in async_step_cloud
    miio_cloud = MiCloud(cloud_username, cloud_password)

2024-09-27 08:23:02.490 WARNING (MainThread) [homeassistant.util.loop] Detected blocking call to open with args ('/etc/sysconfig/clock', 'rt') inside the event loop by integration 'xiaomi_miio' at homeassistant/components/xiaomi_miio/config_flow.py, line 240: miio_cloud = MiCloud(cloud_username, cloud_password) (offender: /home/vscode/.local/ha-venv/lib/python3.12/site-packages/tzlocal/unix.py, line 97: with open(tzpath, "rt") as tzfile:), please create a bug report at https://github.com/home-assistant/core/issues?q=is%3Aopen+is%3Aissue+label%3A%22integration%3A+xiaomi_miio%22
For developers, please see https://developers.home-assistant.io/docs/asyncio_blocking_operations/#open
Traceback (most recent call last):
  File "/home/vscode/.local/ha-venv/bin/hass", line 8, in <module>
    sys.exit(main())
  File "/workspaces/ha-core/homeassistant/__main__.py", line 209, in main
    exit_code = runner.run(runtime_conf)
  File "/workspaces/ha-core/homeassistant/runner.py", line 189, in run
    return loop.run_until_complete(setup_and_run_hass(runtime_config))
  File "/usr/local/lib/python3.12/asyncio/base_events.py", line 674, in run_until_complete
    self.run_forever()
  File "/usr/local/lib/python3.12/asyncio/base_events.py", line 641, in run_forever
    self._run_once()
  File "/usr/local/lib/python3.12/asyncio/base_events.py", line 1979, in _run_once
    handle._run()
  File "/usr/local/lib/python3.12/asyncio/events.py", line 88, in _run
    self._context.run(self._callback, *self._args)
  File "/home/vscode/.local/ha-venv/lib/python3.12/site-packages/aiohttp/web_protocol.py", line 477, in _handle_request
    resp = await request_handler(request)
  File "/home/vscode/.local/ha-venv/lib/python3.12/site-packages/aiohttp/web_app.py", line 559, in _handle
    return await handler(request)
  File "/home/vscode/.local/ha-venv/lib/python3.12/site-packages/aiohttp/web_middlewares.py", line 117, in impl
    return await handler(request)
  File "/workspaces/ha-core/homeassistant/components/http/security_filter.py", line 92, in security_filter_middleware
    return await handler(request)
  File "/workspaces/ha-core/homeassistant/components/http/forwarded.py", line 83, in forwarded_middleware
    return await handler(request)
  File "/workspaces/ha-core/homeassistant/components/http/request_context.py", line 26, in request_context_middleware
    return await handler(request)
  File "/workspaces/ha-core/homeassistant/components/http/ban.py", line 85, in ban_middleware
    return await handler(request)
  File "/workspaces/ha-core/homeassistant/components/http/auth.py", line 242, in auth_middleware
    return await handler(request)
  File "/workspaces/ha-core/homeassistant/components/http/headers.py", line 32, in headers_middleware
    response = await handler(request)
  File "/workspaces/ha-core/homeassistant/helpers/http.py", line 73, in handle
    result = await handler(request, **request.match_info)
  File "/workspaces/ha-core/homeassistant/components/http/decorators.py", line 81, in with_admin
    return await func(self, request, *args, **kwargs)
  File "/workspaces/ha-core/homeassistant/components/config/config_entries.py", line 222, in post
    return await super().post(request, flow_id)
  File "/workspaces/ha-core/homeassistant/components/http/data_validator.py", line 74, in wrapper
    return await method(view, request, data, *args, **kwargs)
  File "/workspaces/ha-core/homeassistant/helpers/data_entry_flow.py", line 122, in post
    result = await self._flow_mgr.async_configure(flow_id, data)
  File "/workspaces/ha-core/homeassistant/data_entry_flow.py", line 370, in async_configure
    result = await self._async_configure(flow_id, user_input)
  File "/workspaces/ha-core/homeassistant/data_entry_flow.py", line 417, in _async_configure
    result = await self._async_handle_step(
  File "/workspaces/ha-core/homeassistant/data_entry_flow.py", line 520, in _async_handle_step
    result: _FlowResultT = await getattr(flow, method)(user_input)
  File "/workspaces/ha-core/homeassistant/components/xiaomi_miio/config_flow.py", line 240, in async_step_cloud
    miio_cloud = MiCloud(cloud_username, cloud_password)

2024-09-27 08:23:02.503 WARNING (MainThread) [homeassistant.util.loop] Detected blocking call to import_module with args ('tzdata.zoneinfo.usr.share.zoneinfo.Etc',) inside the event loop by integration 'xiaomi_miio' at homeassistant/components/xiaomi_miio/config_flow.py, line 240: miio_cloud = MiCloud(cloud_username, cloud_password) (offender: /usr/local/lib/python3.12/importlib/resources/_common.py, line 82: return importlib.import_module(cand)), please create a bug report at https://github.com/home-assistant/core/issues?q=is%3Aopen+is%3Aissue+label%3A%22integration%3A+xiaomi_miio%22
For developers, please see https://developers.home-assistant.io/docs/asyncio_blocking_operations/#import_module
Traceback (most recent call last):
  File "/home/vscode/.local/ha-venv/bin/hass", line 8, in <module>
    sys.exit(main())
  File "/workspaces/ha-core/homeassistant/__main__.py", line 209, in main
    exit_code = runner.run(runtime_conf)
  File "/workspaces/ha-core/homeassistant/runner.py", line 189, in run
    return loop.run_until_complete(setup_and_run_hass(runtime_config))
  File "/usr/local/lib/python3.12/asyncio/base_events.py", line 674, in run_until_complete
    self.run_forever()
  File "/usr/local/lib/python3.12/asyncio/base_events.py", line 641, in run_forever
    self._run_once()
  File "/usr/local/lib/python3.12/asyncio/base_events.py", line 1979, in _run_once
    handle._run()
  File "/usr/local/lib/python3.12/asyncio/events.py", line 88, in _run
    self._context.run(self._callback, *self._args)
  File "/home/vscode/.local/ha-venv/lib/python3.12/site-packages/aiohttp/web_protocol.py", line 477, in _handle_request
    resp = await request_handler(request)
  File "/home/vscode/.local/ha-venv/lib/python3.12/site-packages/aiohttp/web_app.py", line 559, in _handle
    return await handler(request)
  File "/home/vscode/.local/ha-venv/lib/python3.12/site-packages/aiohttp/web_middlewares.py", line 117, in impl
    return await handler(request)
  File "/workspaces/ha-core/homeassistant/components/http/security_filter.py", line 92, in security_filter_middleware
    return await handler(request)
  File "/workspaces/ha-core/homeassistant/components/http/forwarded.py", line 83, in forwarded_middleware
    return await handler(request)
  File "/workspaces/ha-core/homeassistant/components/http/request_context.py", line 26, in request_context_middleware
    return await handler(request)
  File "/workspaces/ha-core/homeassistant/components/http/ban.py", line 85, in ban_middleware
    return await handler(request)
  File "/workspaces/ha-core/homeassistant/components/http/auth.py", line 242, in auth_middleware
    return await handler(request)
  File "/workspaces/ha-core/homeassistant/components/http/headers.py", line 32, in headers_middleware
    response = await handler(request)
  File "/workspaces/ha-core/homeassistant/helpers/http.py", line 73, in handle
    result = await handler(request, **request.match_info)
  File "/workspaces/ha-core/homeassistant/components/http/decorators.py", line 81, in with_admin
    return await func(self, request, *args, **kwargs)
  File "/workspaces/ha-core/homeassistant/components/config/config_entries.py", line 222, in post
    return await super().post(request, flow_id)
  File "/workspaces/ha-core/homeassistant/components/http/data_validator.py", line 74, in wrapper
    return await method(view, request, data, *args, **kwargs)
  File "/workspaces/ha-core/homeassistant/helpers/data_entry_flow.py", line 122, in post
    result = await self._flow_mgr.async_configure(flow_id, data)
  File "/workspaces/ha-core/homeassistant/data_entry_flow.py", line 370, in async_configure
    result = await self._async_configure(flow_id, user_input)
  File "/workspaces/ha-core/homeassistant/data_entry_flow.py", line 417, in _async_configure
    result = await self._async_handle_step(
  File "/workspaces/ha-core/homeassistant/data_entry_flow.py", line 520, in _async_handle_step
    result: _FlowResultT = await getattr(flow, method)(user_input)
  File "/workspaces/ha-core/homeassistant/components/xiaomi_miio/config_flow.py", line 240, in async_step_cloud
    miio_cloud = MiCloud(cloud_username, cloud_password)
```

Reply:
Wow! Good catch & thanks for the details!